lib/puppet/defaults.rb

```diff
@@ -32,6 +32,41 @@ def self.valid_file_checksum_types
       %w[sha256 sha256lite sha384 sha512 sha224 sha1 sha1lite md5 md5lite mtime ctime]
   end
 
+  def self.log_ca_migration_warning
+    urge_to_migrate = <<-UTM
+The cadir is currently in your puppet/ssl directory. This location will be
+deprecated and removed in a future version of puppet. Please run the the
+puppetserver ca tool to migrate it out to the puppetserver directory. Use
+`puppetserver ca migrate --help` for more info.
+UTM
+    Puppet.warn_once('deprecations',
```

Comment:
I used `warn_once` here because the only hook that worked was `:on_initialize_and_write`; other hooks always ran into interpolation problems with `$ssldir` and `$confdir`. Using `:on_initialize_and_write` ended up emitting a warning 2-3 times.

@joshcooper is that the right solution in this case? I'm not completely clear on why it emits so much...or maybe we want it to emit a lot to nag users?

Reply:
yeah... these hooks are the worst. Hooks are sometimes called with the uninterpolated value, sometimes with the interpolated value and sometimes with a string. And each hook type is called at different points in the application lifecycle.

In general `on_initialize_and_write` hooks are the ones to use. Otherwise if a setting is defined in `[agent]` or `[server]` then the value will take effect, but the  hook will never be called! Using `warn_once` also makes sense. I think you'll find that the first couple of times the hook is called, attempts to resolve `Puppet[:ssldir]` will [raise an InterpolationError, which is silently rescued](https://github.com/puppetlabs/puppet/blob/d61cf32faae73454d4b0dcc2f199466a457e8397/lib/puppet/settings.rb#L428), and settings code calls the hook again later.

Gabi and I have been talking about eliminating `on_write_only` hooks, and possibly only calling the hook once the value has been resolved/interpolated. Or perhaps calling the hook with an event name, so it's clear why it's being called, e.g. `:on_default`, `:on_global_initialized`, `:on_application_initialized`, `:on_changed`